### PR TITLE
Fix checkpoint callback & Trainer.test(_) issue for TPUs

### LINKED ISSUE #6230: TPU:  Crashes using trainer.test()
# 🐛 Bug

`trainer.test()` does not work with TPUs.

There are a few different ways we've seen it crash.

### 1. Looks like a call to `barrier()` coming from `__test_using_best_weights`
```
RuntimeError                              Traceback (most recent call last)
<ipython-input-17-587e2a9e3858> in <module>
----> 1 trainer.test(datamodule=dm)

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in test(self, model, test_dataloaders, ckpt_path, verbose, datamodule)
    922             results = self.__test_given_model(model, test_dataloaders)
    923         else:
--> 924             results = self.__test_using_best_weights(ckpt_path, test_dataloaders)
    925 
    926         self.teardown('test')

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in __test_using_best_weights(self, ckpt_path, test_dataloaders)
    950                 return {}
    951             if not self._device_type == DeviceType.TPU:
--> 952                 self.accelerator.barrier()
    953 
    954             ckpt = pl_load(ckpt_path, map_location=lambda storage, loc: storage)

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py in barrier(self, name)
    375 
    376     def barrier(self, name: Optional[str] = None) -> None:
--> 377         self.training_type_plugin.barrier(name=name)
    378 
    379     def broadcast(self, obj: object, src: int = 0) -> object:

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py in barrier(self, name)
    113 
    114     def barrier(self, name: Optional[str] = None) -> None:
--> 115         rendezvous(f"pl.Trainer.{name}")
    116 
    117     def transfer_distrib_spawn_state_on_fit_end(self, results):

/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py in rendezvous(tag, payload, replicas)
    859     ordinal `i` at position `i` in the returned tuple.
    860   """
--> 861   return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
    862 
    863 

RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:316 : Check failed: impl_->channel->WaitForConnected( std::chrono::system_clock::now() + std::chrono::seconds(connect_wait_seconds))  
```
So the barrier is coming from [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L915). This is strange that `barrier` is being called - I think this means that `if not self._device_type == DeviceType.TPU` is mistakenly evaluating to `True`?  I think pytorch lightning spins up 8 processes for 8 TPU cores, is it possible only some of them are evaluating to `True`?

Basically it seems like at least 1 process is not making it to this point, which means the other processes are waiting in the barrier and the meetup never happens so we get the `RuntimeError` shown.

### 2. Looks like a call to `xm.save()` is being misused:
```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 103, in new_process
    self.transfer_distrib_spawn_state_on_fit_end(results)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 129, in transfer_distrib_spawn_state_on_fit_end
    xm.save(self.lightning_module.state_dict(), last_path)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 817, in save
    rendezvous('torch_xla.core.xla_model.save')
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 861, in rendezvous
    return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 103, in new_process
    self.transfer_distrib_spawn_state_on_fit_end(results)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 129, in transfer_distrib_spawn_state_on_fit_end
    xm.save(self.lightning_module.state_dict(), last_path)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 817, in save
    rendezvous('torch_xla.core.xla_model.save')
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 861, in rendezvous
    return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
Exception in device=TPU:6: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)Exception in device=TPU:3: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)
```
I think the problem is [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/tpu_spawn.py#L140) with the usage of `xm.save()`.

`xm.save()` already handles the multiprocess case by checking the ordinal and only writing to disk if the process is on the master ordinal.  In general, if you surround `xm.save()` with `if` statements, it means some TPU cores enter the `if` statement and some will not, so the cores that entered the `if` statement will be waiting for those that didn't enter and eventually it will time out and crash.

# Repro methods

### 1. (Colab) Make 3 modifications to the  [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)
1. Switch runtime version to TPU
2. Add this cell as the first cell:
```
VERSION = "1.7"  #@param ["1.7" , "20200516", "nightly"]
!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py --version $VERSION
```
3. Add `tpu_cores=8` to the trainer cell

### 2. (Google Cloud) Use the attached repro.py file in the following way:
1. create a TPU
2. create a Google Cloud VM (I used e2-standard-32 but size shouldn't matter too much)
3. SSH into VM
4. (VM) `conda activate torch-xla-1.7`
5. (VM) `pip install pytorch-lightning==1.2.1`
5. (VM) `export TPU_IP_ADDRESS=my.tpu.ip.addr`
6. (VM) `export XRT_TPU_CONFIG="tpu_worker;0;$TPU_IP_ADDRESS:8470"`
7. (VM) `python3 repro.py`

### 3. (Your CI setup) Modify TPU unit tests as follows:
* Add a `trainer.test(test_dataloaders=DataLoader(RandomDataset(32, 2000), batch_size=32))` after some call to `trainer.fit`
* For example, I changed test_model_tpu_early_stop test to look like this:
```
@pytest.mark.skipif(not _TPU_AVAILABLE, reason="test requires TPU machine")
@pl_multi_process_test
def test_model_tpu_early_stop(tmpdir):
    """Test if single TPU core training works"""

    # todo: Test on 8 cores - hanging.

    class CustomBoringModel(BoringModel):

        def validation_step(self, *args, **kwargs):
            out = super().validation_step(*args, **kwargs)
            self.log('val_loss', out['x'])
            return out

    tutils.reset_seed()
    model = CustomBoringModel()
    trainer = Trainer(
        callbacks=[EarlyStopping(monitor='val_loss')],
        default_root_dir=tmpdir,
        progress_bar_refresh_rate=0,
        max_epochs=2,
        limit_train_batches=2,
        limit_val_batches=2,
        tpu_cores=[1],
    )
    trainer.fit(model)
+    trainer.test(test_dataloaders=DataLoader(RandomDataset(32, 2000), batch_size=32))
```

Ran tests with `coverage run --source=pytorch_lightning -m pytest tests/models/test_tpu.py -v`. This should allow testing on the CI framework

### Environment

 - PyTorch Version (e.g., 1.0): 1.7
 - OS (e.g., Linux): Linux
 - Build command you used (if compiling from source): `pip install pytorch-lightning==1.2.1` (note that earlier versions hang due to #5841 )
 - Python version: 3.6
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fix checkpoint callback issue for TPUs when set False & `trainer.test(..).
Fixes #6230 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/test_tpu.py [hunk 1]
```diff
@@ -383,3 +383,18 @@ def test_tpu_precision_16_clip_gradients(mock_clip_grad_norm, clip_val, tmpdir):
...
         mock_clip_grad_norm.assert_not_called()
+
+
+@RunIf(tpu=True)
+@pl_multi_process_test
+def test_if_test_works_with_checkpoint_false(tmpdir):
+    """
+    Ensure that model trains properly when
+    `checkpoint_callback` is set to False.
+    """
```
Comment (carmocca) [inline tests/models/test_tpu.py:394]: nit: this fits into 120 characters

```suggestion
    """Ensure that model trains properly when `checkpoint_callback` is set to False."""
```

File: tests/models/test_tpu.py [hunk 2]
```diff
@@ -383,3 +383,18 @@ def test_tpu_precision_16_clip_gradients(mock_clip_grad_norm, clip_val, tmpdir):
         mock_clip_grad_norm.assert_called()
     else:
         mock_clip_grad_norm.assert_not_called()
+
+
+@RunIf(tpu=True)
+@pl_multi_process_test
```
Comment (carmocca) [inline tests/models/test_tpu.py:389]: noob question: what does this do?

File: CHANGELOG.md [hunk 3]
```diff
@@ -168,6 +168,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed comparing required versions ([#6434](https://github.com/PyTorchLightning/pytorch-lightning/pull/6434))
 
 
+- Fixed checkpoint callback issue with TPUs when set False ([#6654](https://github.com/PyTorchLightning/pytorch-lightning/pull/6654))
```
Comment (carmocca) [inline CHANGELOG.md:171]:
```suggestion
- Fixed error on TPUs when there was no `ModelCheckpoint` ([#6654](https://github.com/PyTorchLightning/pytorch-lightning/pull/6654))
```

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 4]
```diff
@@ -130,7 +130,8 @@ def barrier(self, name: Optional[str] = None) -> None:
             rendezvous(f"pl.Trainer.{name}")
 
     def transfer_distrib_spawn_state_on_fit_end(self, results):
-        best_model_path = self.lightning_module.trainer.checkpoint_callback.best_model_path
+        checkpoint_callback = self.lightning_module.trainer.checkpoint_callback
+        best_model_path = checkpoint_callback.best_model_path if checkpoint_callback else None
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:134]: what happens if there are multiple checkpoint callbacks attached?

File: tests/models/test_tpu.py (hunk 2 above)
Comment (tchaton) [reply tests/models/test_tpu.py:389]: TPU attach the tpu_cores to the process. pl_multi_process_test Popen the function into a subprocess, so pytest doesn't hang as tpu_cores would have already been assigned.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:134]: I think I'd rather use the first `best` path and log the path used when running `test`

**Review by SeanNaren — APPROVED** [review]
Nice! on a side note how far are we from enabling TPU tests again?

**Comment by kaushikb11** [comment]
@SeanNaren Thanks for bringing it up! We could keep a look on it for next couple of days, I have seen instances when the TPU Pods used to get killed mid testing. Afk, will create an Issue and track it.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 5]
```diff
@@ -109,13 +108,15 @@ def new_process(self, process_idx: int, trainer, mp_queue) -> None:
 
         # replace trainer save_checkpoint to use `xm.save`
         trainer.save_checkpoint = self.save_checkpoint
-        self.barrier()
+        self.barrier("pre run stage")
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:111]:
```suggestion
        self.barrier("pre-run-stage")
```
space can be tricky

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 6]
```diff
@@ -109,13 +108,15 @@ def new_process(self, process_idx: int, trainer, mp_queue) -> None:
...
 
         results = trainer.run_stage()
 
         self.__save_end_of_training_weights(self.lightning_module)
         self.transfer_distrib_spawn_state_on_fit_end(results)
 
+        self.barrier("end process")
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:118]:
```suggestion
        self.barrier("end-process")
```

**Review by Borda — COMMENTED** [review]
have you tested it locally as TPU is put for now...?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py]: Done!

**Comment by kaushikb11** [comment]
@Borda 
Screen Shot 2021-03-25 at 1 53 13 PM

All tests are passing. Had to do a tweek. Will investigate it soon.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 7]
```diff
@@ -109,13 +108,15 @@ def new_process(self, process_idx: int, trainer, mp_queue) -> None:
...
         trainer.save_checkpoint = self.save_checkpoint
-        self.barrier()
+        self.barrier("pre-run-stage")
 
         results = trainer.run_stage()
 
         self.__save_end_of_training_weights(self.lightning_module)
         self.transfer_distrib_spawn_state_on_fit_end(results)
 
+        self.barrier("end-process")
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:118]: is this needed?

File: tests/models/test_tpu.py [hunk 8]
```diff
@@ -357,13 +357,14 @@ def test_reduce(rank):
     xmp.spawn(test_reduce, nprocs=8, start_method='fork')
 
 
-@pytest.mark.parametrize("clip_val", [0, 10])
+@pytest.mark.parametrize("clip_val", [10])
 @RunIf(tpu=True)
 @pl_multi_process_test
 @mock.patch("pytorch_lightning.accelerators.tpu.xla_clip_grad_norm_")
```
Comment (Borda) [inline tests/models/test_tpu.py:363]:
```suggestion
@RunIf(tpu=True)
@pl_multi_process_test
@pytest.mark.parametrize("clip_val", [10])
@mock.patch("pytorch_lightning.accelerators.tpu.xla_clip_grad_norm_")
```

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 7 above)
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:118]: Oh yes! We need a `rendezvous` before the end of the process.

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -983,7 +983,8 @@ def __load_ckpt_weights(
                     ' specify a path for a checkpoint `.{fn}(ckpt_path=PATH)`'
                 )
 
-            self.training_type_plugin.barrier()
+            if not self._device_type == DeviceType.TPU:
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:986]: Can you explain why is it needed ?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:986]: Please, add a comment to explain why it is needed.

Comment (kaushikb11) [reply pytorch_lightning/trainer/trainer.py:987]: Comment added. Thanks.

**Pull request merged**